File: problem-set-6-sentimental-credit/credit.py
import re

def main():
    # Get credit card number from user
    card_number = input("Number: ")

    # Validate credit card number
    if is_valid_card(card_number):
        print_card_type(card_number)
    else:
        print("INVALID")

def is_valid_card(card_number):
    # Check if the card number consists of only digits and has a valid length
    if not re.match("^\d+$", card_number) or len(card_number) < 13 or len(card_number) > 16:
        return False

    # Reverse the card number and double every second digit
    reversed_card = card_number[::-1]
    doubled_digits = [int(reversed_card[i]) * 2 if i % 2 == 1 else int(reversed_card[i]) for i in range(len(reversed_card))]

    # Sum up the digits of the doubled numbers and the unaffected numbers
    total = sum([(n // 10) + (n % 10) for n in doubled_digits])

    # Check if the total modulo 10 equals 0
    return total % 10 == 0

def print_card_type(card_number):
    # Determine the card type based on the first few digits of the card number
    if re.match("^3[47]", card_number) and len(card_number) == 15:
        print("AMEX")
    elif re.match("^5[1-5]", card_number) and len(card_number) == 16:
        print("MASTERCARD")
    elif re.match("^4", card_number) and (len(card_number) == 13 or len(card_number) == 16):
        print("VISA")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()

File: problem-set-1-credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    // Prompt the user for a credit card number
    long long card_number;
    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number <= 0);

    // Validate the credit card number using Luhn's Algorithm
    int sum = 0;
    int digit_count = 0;
    bool is_second_digit = false;
    while (card_number > 0)
    {
        int digit = card_number % 10;
        if (is_second_digit)
        {
            digit *= 2;
            while (digit > 0)
            {
                sum += digit % 10;
                digit /= 10;
            }
        }
        else
        {
            sum += digit;
        }
        card_number /= 10;
        is_second_digit = !is_second_digit;
        digit_count++;
    }

    // Check if the sum's last digit is 0
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Determine the type of credit card
    if ((digit_count == 15) && (card_number == 34 || card_number == 37))
    {
        printf("AMEX\n");
    }
    else if ((digit_count == 16) && (card_number >= 51 && card_number <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((digit_count == 13 || digit_count == 16) && (card_number / 10 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}
